articles/113_ros_time.md

```diff
@@ -0,0 +1,154 @@
+When the external time source is active `ROSTime` will return the latest value reported by the Time Source.
+
+### Time Sources
+
+#### Default Time Source
+
+To implement the time abstraction the following approach will be used.
+
+The time abstraction can be published by one source on the `/clock` topic.
+The topic will contain the most up to date time for the ROS system.
+If a publisher exists for the topic, it will override the system time when using the ROS time abstraction.
+If `/clock` is being published, calls to the ROS time abstraction will return the latest time received from the `/clock` topic.
+If time has not been set it will return zero if nothing has been received.
+A time value of zero should be considered an error meaning that time is uninitialized.
+
+If the time on the clock jumps backwards, a callback handler will be invoked and be required to complete before any calls to the ROS time abstraction report the new time.
+Calls that come in before that must block.
+The developer has the opportunity to register callbacks with the handler to clear any state from their system if necessary before time will be in the past.
+
+The frequency of publishing the `/clock` as well as the granularity are not specified as they are application specific.
+
+##### No Advanced Estimating Clock
+
+There are more advanced techniques which could be included to attempt to estimate the propagation properties and extrapolate between time ticks.
+However all of these techniques will require making assumptions about the future behavior of the time abstraction.
+And in the case that playback or simulation is instantaneously paused, it will break any of these assumptions.
+There are techniques which would allow potential interpolation, however to make these possible it would require providing guarantees about the continuity of time into the future.
+For more accuracy the progress of time can be slowed, or the frequency of publishing can be increased.
+Tuning the parameters for the `/clock` topic lets you trade off time for computational effort and/or bandwidth.
+
+#### Custom Time Source
+
+It is possible that the user may have access to an out of band time source which can provide better performance than the default source the `/clock` topic.
+A potential alternative source of time might be a GPS.
+The user will be able to switch out the time source for either each instance of their Time object as well as the ability to override the default for the process.
+
+## Implementation
+
+The public API will be provided by the client libraries.
+The client libraries will be able to rely on a base implementation in rcl which provides a common behavior.
+If the client library implementer chooses they my also reimplement the core behavior, however it is expected that the rcl implementation should be adaquate to support any language with a foreign function interface for C code.
```

Comment:
The first sentence is a bit ambiguous to me, maybe instead replace the first two sentences with:

> The SystemTime, SteadyTime, and ROSTime API's will be provided by each client library in an idiomatic way, but they may share a common implementation, e.g. provided by `rcl`.

Then shorten the last sentence just to make it clear that it is not optional:

> However, if a client library chooses to not use the shared implementation then it must implement the functionality itself.